Framework/Core/include/Framework/MessageSet.h

```diff
@@ -23,22 +23,28 @@ namespace framework
 
 /// A set of associated inflight messages.
 struct MessageSet {
-  std::vector<PartRef> parts;
+  struct Index {
+    Index(size_t p, size_t s) : position(p), size(s) {}
+    size_t position = 0;
+    size_t size = 0;
+  };
+  std::vector<FairMQMessagePtr> messages;
+  std::vector<Index> index;
```

Comment:
Can you remind why we need the index at all?

Reply:
this will contain the length of the message train, so we have the position in the linear vector and the number of messages for a particular part (header plus n payloads)